lxddoc: a go-swagger like documentation tool (#11652)
lxc/lxd: .sphinx/lxd-doc/README.md

# lxd-doc

A small CLI to parse comments in a Golang codebase meant to be used for a documentation tool (like Sphinx for example).

It parses the comments from the AST and extracts their documentation.

# Usage

    $ lxd-doc -h
    Usage of lxd-doc:
    -e value
        Path that will be excluded from the process

# Formatting

A comment is formatted this way:

```go
	// lxddoc:generate(group=cluster, key=scheduler.instance)
	//
	//  <Possibly a very long documentation on multiple lines with Markdown tables, etc.>
	// ---
	//  shortdesc: Possible values are all, manual and group. See Automatic placement of instances for more information.
    //  condition: container
	//  default: `all`
	//  type: integer

> 	//  liveupdate: `yes`

Reviewer: Now the <ANY_KEY>: <ANY_VALUE> line is missing?
(And the indentation is messed up again :wink: )

Author: @ru-fu somehow, on my side, the `<ANY_KEY>: <ANY_VALUE>` is pproperly indented but on github it is different...

Author: Ok I managed :) (vscode tab config was weird)

Reviewer: Thanks, all good now! :)

> # lxddoc

Reviewer: ```suggestion
# lxd-doc
```

A small CLI to parse comments in a Golang codebase meant to be used for a documentation tool (like Sphinx for example).

It parses the comments from the AST and extracts their documentation.

# Usage

    $ lxd-doc -h
    Usage of lxd-doc:
    -e value
        Path that will be excluded from the process

# Formatting

A comment is formatted this way:

```go
	// lxddoc:generate(group=cluster, key=scheduler.instance)
	//
	//  <Possibly a very long documentation on multiple lines with Markdown tables, etc.>
	// ---

> 	//  shortdesc: Possible values are all, manual and group. See Automatic placement of instances for more information.
> 	//  <ANY_KEY>: <ANY_VALUE>
>     //  condition: container
> 	//  default: `all`
> 	//  type: integer
> 	//  liveupdate: `yes`

Reviewer: ```suggestion
	//  shortdesc: Possible values are all, manual and group. See Automatic placement of instances for more information.
        //  condition: container
	//  default: `all`
	//  type: integer
	//  liveupdate: `yes`
	//  <ANY_KEY>: <ANY_VALUE>
```

We should have all of them at the same indentation. ;)
And imo it makes most sense to have the "any key" one at the end.

		//
		//  ---
		//	shortdesc: Free form user key/value storage (can be used in search).
		//	condition: container
		//	default: -
		//	type: string
		//	liveupdate: `yes`
		if strings.HasPrefix(k, "user.") {
			continue
		}

		validator, ok := clusterConfigKeys[k]
		if !ok {
			return fmt.Errorf("Invalid cluster configuration key %q", k)
		}

		err := validator(v)
		if err != nil {
			return fmt.Errorf("Invalid cluster configuration key %q value", k)
		}
	}

	return nil
```

The go-swagger spec from source generator can only handles `swagger:meta` (global file/package level documentation), `swagger:route` (API endpoints), `swagger:params` (function parameters), `swagger:operation` (method documentation), `swagger:response` (API response content documentation), `swagger:model` (struct documentation) generation. In our use case, we would want a config variable spec generator that can bundle any key-value data pairs alongside metadata to build a sense of hierarchy and identity (we want to associate a unique key to each lxddoc comment group that will also be displayed in the generated documentation)

> In a swagger fashion, `lxddoc` can associate metadata key-value pairs (here for example, `group` and `key`) to data key-value pairs. As a result, it can generate a YAML tree out of the code documentation and also a Markdown document.

Reviewer: ```suggestion
In a swagger fashion, `lxd-doc` can associate metadata key-value pairs (here for example, `group` and `key`) to data key-value pairs. As a result, it can generate a YAML tree out of the code documentation and also a Markdown document.
```
